### ISSUE TITLE
Longer `GRPC_TIMEOUT` for larger upload support

### ISSUE THREAD

**Issue by ackagel** [vanvalenlab/kiosk-redis-consumer#152]
For some larger .zip uploads (>50MB) on deepcell.org, a random selection of the images aren't processed and show a generic version of this error:

```
Traceback (most recent call last):
File "/usr/src/app/redis_consumer/consumers/base_consumer.py", line 202, in consume
status = self._consume(redis_hash)
File "/usr/src/app/redis_consumer/consumers/multiplex_consumer.py", line 124, in _consume
image = self.predict(image, model_name, model_version)
File "/usr/src/app/redis_consumer/consumers/base_consumer.py", line 614, in predict
model_dtype, untile=untile)
File "/usr/src/app/redis_consumer/consumers/base_consumer.py", line 464, in _predict_big_image
in_tensor_dtype=model_dtype)
File "/usr/src/app/redis_consumer/consumers/base_consumer.py", line 311, in grpc_image
prediction = client.predict(req_data, settings.GRPC_TIMEOUT)
File "/usr/src/app/redis_consumer/grpc_clients.py", line 185, in predict
response = self._retry_grpc(request, request_timeout)
File "/usr/src/app/redis_consumer/grpc_clients.py", line 134, in _retry_grpc
raise err
File "/usr/src/app/redis_consumer/grpc_clients.py", line 123, in _retry_grpc
response = api_call(request, timeout=request_timeout)
File "/usr/local/lib/python3.6/site-packages/grpc/_channel.py", line 826, in __call__
return _end_unary_response_blocking(state, call, False, None)
File "/usr/local/lib/python3.6/site-packages/grpc/_channel.py", line 729, in _end_unary_response_blocking
raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
status = StatusCode.DEADLINE_EXCEEDED
details = "Deadline Exceeded"
debug_error_string = "{"created":"@1609789741.491197858","description":"Error received from peer ipv4:10.43.253.0:8500","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Deadline Exceeded","grpc_status":4}"
```

Breaking a larger .zip into multiple smaller .zip's fixes this, which makes me suspect that each 'server' is taking longer having to deal with multiple files to segment/process, which ends up lasting longer than the `GRPC_TIMEOUT`.

If I haven't misunderstood the problem, could the `GRPC_TIMEOUT` env var be set to something like 60s instead of 30s?

**Comment by ngreenwald** [comment]
@willgraf, any suggestions on this?

**Comment by willgraf** [comment]
I'll try adjusting `GRPC_TIMEOUT` to 60s to see if that helps the issue.

**Comment by ngreenwald** [comment]
Great, @ackagel let us know if this solves it

**Comment by ackagel** [comment]
Looks like extending `GRPC_TIMEOUT` didn't help; got the same error messages.

**Comment by willgraf** [comment]
The deepcell.org instance is a small cluster intended for demo purposes rather than large-scale processing. Each file can be retried a few times (maybe 5?) but then it fails to prevent jobs from being infinitely retried.

Perhaps we could treat the `DEADLINE_EXCEEDED` error as "not a true failure" like [we do with `UNAVAILABLE`](https://github.com/vanvalenlab/kiosk-redis-consumer/blob/cefc4736d1658505966bd8bbe3638a67b83290f2/redis_consumer/grpc_clients.py#L138).  This would allow the job to be retried indefinitely, but may have some unexpected consequences.

In the meantime, I recommend using smaller zip files or using the CLI to see if that resolves anything.

**Comment by ngreenwald** [comment]
We're currently testing this with a zip file with 1,000 copies of the same image. Is there something about the total number of files in the zip that is causing the problem? Each individual test image runs fine. e.g. is the timeout related to the total amount of time spent on the entire zip, rather than the amount spend on each individual image?

**Comment by willgraf** [comment]
Each of the 1k images is cropped into the model input shape and each crop is sent over as a request to TensorFlow Serving. The `DEADLINE_EXCEEDED` error is thrown when TensorFlow Serving has not responded to a request in `GRPC_TIMEOUT` seconds.

This usually occurs when there are too many jobs being processed at once, so each job slows down. We are batching requests to prevent this, though the batch sizes were configured using a different model. I've just now adjusted the `TF_MAX_BATCH_SIZE` from 32 to 8 - let's see if that helps.

**Comment by ackagel** [comment]
*375 copies (bad upload speed on my end made the 1k upload too long).  Okay great, I'll test that out, thanks!

**Comment by ackagel** [comment]
looks like that largely did the trick, but I did get a couple instances of this new error (for 5-10 of the last few tifs in the upload):

```
multiplex:fov8_365.tif:322075af4caa4cacab46a0f7e82f5986=Traceback (most recent call last):
  File "/usr/src/app/redis_consumer/consumers/base_consumer.py", line 202, in consume
    status = self._consume(redis_hash)
  File "/usr/src/app/redis_consumer/consumers/multiplex_consumer.py", line 76, in _consume
    fname = self.storage.download(hvals.get('input_file_name'), tempdir)
  File "/usr/src/app/redis_consumer/storage.py", line 278, in download
    dest = self.get_download_path(filepath, download_dir)
  File "/usr/src/app/redis_consumer/storage.py", line 115, in get_download_path
    no_upload_dir = os.path.join(*(filepath.split(os.path.sep)[1:]))
AttributeError: 'NoneType' object has no attribute 'split'
```

looks like the 'input_file_name' key isn't found in hvals or is still `None`

**Comment by willgraf** [comment]
That is very strange! Before we start to download the file we check that the job hash is valid, which inspects the `input_file_name` key. Something happened between the `is_valid_hash` call and the `hgetall` call that removed the data from Redis. Maybe (somehow) the key was worked on by 2 different consumers, and the key expired between the two calls?

Can you try to replicate the error, and if you can, could you please make a new issue for it?

**Comment by willgraf** [comment]
Closing this as it seems the issue was setting `TF_MAX_BATCH_SIZE` properly for new models.
